File: examples/scala/ConnectViewOnlyClientExample.scala
// Connect to a networked openDAQ device as a view-only client (reads work,
// configuration writes are refused).  Expects a device/simulator at
// daq.nd://127.0.0.1 or at $OPENDAQ_DEVICE.

import com.opendaq.{Unit as _, *}   // hide com.opendaq.Unit so `: Unit` means scala.Unit

object ConnectViewOnlyClientExample:
  def main(args: Array[String]): Unit =
    val instance = Instance()

    val config = instance.createDefaultAddDeviceConfig()
    val general = config.getPropertyValue("General").asInstanceOf[DaqObject].asType(classOf[PropertyObject])
    val clientType = general.getProperty("ClientType")
    val options = Daq.unbox(clientType.getSelectionValues)
    println("Available ClientType options:")
    options match
      case map: java.util.Map[?, ?] =>
        val sorted = new java.util.TreeMap[java.lang.Long, AnyRef]()
        map.entrySet.forEach { e =>
          sorted.put(e.getKey.asInstanceOf[Number].longValue, e.getValue.asInstanceOf[AnyRef])
        }
        sorted.forEach((key, value) => println("  " + key + " = " + value))
      case list: java.util.List[?] =>
        for i <- 0 until list.size do println("  " + i + " = " + list.get(i))
      case _ =>

    config.setPropertyValue("General.ClientType", 2)   // 2 = view-only
    val connectionString = Option(System.getenv("OPENDAQ_DEVICE")).getOrElse("daq.nd://127.0.0.1")

    try
      val device = instance.addDevice(connectionString, config)
      println("Connected to " + device.getName + " as a view-only client.")
      println("Visible signals: " + device.getSignalsRecursive.size)
      val propertyName = firstWritablePropertyName(device)
      if propertyName == null then
        println("Device exposes no writable property to probe.")
      else
        try
          device.setPropertyValue(propertyName, device.getPropertyValue(propertyName))
          println("Unexpected: writing \"" + propertyName + "\" was allowed.")
        catch
          case e: RuntimeException =>
            println("Write to \"" + propertyName + "\" refused, as expected for view-only: " + e.getMessage)
    catch
      case e: RuntimeException =>
        println("Could not connect to " + connectionString + ": " + e.getMessage)
        println("Start an openDAQ device/simulator there, or set OPENDAQ_DEVICE.")

  /**
   * Name of the object's first visible, non-read-only, non-callable property,
   * or null.  Used to probe whether the connection actually refuses writes.
   */
  private def firstWritablePropertyName(obj: PropertyObject): String =
    obj.getVisibleProperties.stream
      .filter(p => !p.getReadOnly && p.getValueType != CoreType.FUNC && p.getValueType != CoreType.PROC)
      .map(_.getName)
      .findFirst
      .orElse(null)

File: examples/ConnectViewOnlyClientExample.java
import java.util.Map;
import java.util.TreeMap;

import com.opendaq.*;

/**
 * Connect to a networked openDAQ device as a view-only client (reads work,
 * configuration writes are refused).  Expects a device/simulator at
 * daq.nd://127.0.0.1 or at $OPENDAQ_DEVICE.
 */
public class ConnectViewOnlyClientExample {

    public static void main(String[] args) {
        Instance instance = new Instance();

        PropertyObject config = instance.createDefaultAddDeviceConfig();

        PropertyObject general = ((DaqObject) config.getPropertyValue("General"))
            .asType(PropertyObject.class);
        Property clientType = general.getProperty("ClientType");
        Object options = Daq.unbox(clientType.getSelectionValues());
        System.out.println("Available ClientType options:");
        if (options instanceof Map<?, ?> map) {
            new TreeMap<>(map.entrySet().stream()
                    .collect(java.util.stream.Collectors.toMap(
                        e -> ((Number) e.getKey()).longValue(), Map.Entry::getValue)))
                .forEach((key, value) -> System.out.println("  " + key + " = " + value));
        } else if (options instanceof java.util.List<?> list) {
            for (int i = 0; i < list.size(); i++) {
                System.out.println("  " + i + " = " + list.get(i));
            }
        }

        config.setPropertyValue("General.ClientType", 2);   // 2 = view-only
        String connectionString = System.getenv().getOrDefault("OPENDAQ_DEVICE", "daq.nd://127.0.0.1");

        try {
            Device device = instance.addDevice(connectionString, config);
            System.out.println("Connected to " + device.getName() + " as a view-only client.");
            System.out.println("Visible signals: " + device.getSignalsRecursive().size());
            String propertyName = firstWritablePropertyName(device);
            if (propertyName == null) {
                System.out.println("Device exposes no writable property to probe.");
                return;
            }
            try {
                device.setPropertyValue(propertyName, device.getPropertyValue(propertyName));
                System.out.println("Unexpected: writing \"" + propertyName + "\" was allowed.");
            } catch (RuntimeException e) {
                System.out.println("Write to \"" + propertyName + "\" refused, as expected for view-only: "
                    + e.getMessage());
            }
        } catch (RuntimeException e) {
            System.out.println("Could not connect to " + connectionString + ": " + e.getMessage());
            System.out.println("Start an openDAQ device/simulator there, or set OPENDAQ_DEVICE.");
        }
    }

    /**
     * Name of the object's first visible, non-read-only, non-callable
     * property, or null.  Used to probe whether the connection actually
     * refuses writes.
     */
    private static String firstWritablePropertyName(PropertyObject object) {
        for (Property property : object.getVisibleProperties()) {
            CoreType type = property.getValueType();
            if (!property.getReadOnly() && type != CoreType.FUNC && type != CoreType.PROC) {
                return property.getName();
            }
        }
        return null;
    }
}
